File: api/app/schemas/category.py
from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, ConfigDict

from .part import PublicPartResponse
from .sponsor import SponsorResponse
from .supplier import SupplierResponse


class FeaturedSupplier(BaseModel):
    """A Featured supplier on a category (id + name).

    Carries the supplier id alongside the name so the admin categories tree can
    unambiguously identify the sponsor — Supplier.name has no unique constraint,
    so two distinct suppliers can share a name and collide in a name-keyed
    lookup. (Featured status is sourced from the `sponsors` table as of
    2026-06-03; the old CategorySupplier feature flag is gone.)
    """

    id: UUID
    name: str
    model_config = ConfigDict(from_attributes=True)


class SubcategoryResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    icon: str
    parts_count: int = 0
    featured_supplier_name: str | None = None
    # All Featured CategorySuppliers for this category, ordered by rank ASC
    # (lowest rank = highest priority). 2026-06-02: a category may have many
    # Featured suppliers; `featured_supplier_name` is kept for back-compat
    # and mirrors `featured_suppliers[0].name`.
    featured_suppliers: list[FeaturedSupplier] = []
    model_config = ConfigDict(from_attributes=True)


class CategoryResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    icon: str
    description: str | None = None
    parts_count: int = 0
    children: list[SubcategoryResponse]
    featured_supplier_name: str | None = None
    featured_suppliers: list[FeaturedSupplier] = []
    model_config = ConfigDict(from_attributes=True)


class ParentCategoryResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    icon: str
    # Sibling subcategories — surfaced so the frontend can render the
    # intra-category chips on a subcategory page (current page marked active)
    # without an extra round-trip to fetch the parent. See 2026-05-16 bug fix
    # in CategoryPage/SubcategoryChips.
    children: list[SubcategoryResponse] = []
    model_config = ConfigDict(from_attributes=True)


class PartsPage(BaseModel):
    items: list[PublicPartResponse] = []
    total: int = 0
    page: int = 1
    pages: int = 1
    per_page: int = 20
    model_config = ConfigDict(from_attributes=True)


PopularPartsPage = PartsPage


class CategoryDetailResponse(CategoryResponse):
    parent: ParentCategoryResponse | None = None
    # `sponsor` = this child's newest visible GOLD sponsor (the single
    # Subcategory Sponsor slot → SponsorBlock). `silver` = this child's visible
    # SILVER sponsors (the multi-occupant directory → SilverPartners). Both are
    # tier-filtered in category_service. (2026-06-11 tier-boards matrix.)
    sponsor: SponsorResponse | None
    silver: list[SupplierResponse] = []
    parts: PartsPage = PartsPage()
    popular_parts: PartsPage = PartsPage()
    model_config = ConfigDict(from_attributes=True)


CategoryDetailResponse.model_rebuild()


class CategoryPartnersResponse(BaseModel):
    """Category Sponsor (Platinum) board payload for a TOP-LEVEL category (split
    out of the heavy CategoryDetailResponse 2026-06-04 so it's a small,
    cacheable, top-level artifact). `slug`/`name` are the RESOLVED top-level
    category — a child slug resolves to its parent — so the same Platinum board
    shows on every subpage. `platinum` is the single visible Platinum sponsor
    (newest-wins); `None` → the board renders its Open-Placement state.
    (2026-06-11 tier-boards matrix — was a multi-supplier `partners` list.)"""

    slug: str
    name: str
    platinum: SponsorResponse | None = None
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/public/types/category.ts
export interface Subcategory {
  id: string;
  name: string;
  slug: string;
  icon: string;
  // Server sends it on category detail children; optional because older
  // cached shells may predate it. `| null` per the ?:-misses-null gotcha.
  parts_count?: number | null;
}

export interface Category {
  id: string;
  name: string;
  slug: string;
  icon: string;
  description?: string | null;
  // OWN-count only — parts attach to subcategories, so a top-level card's
  // real total is this plus the children's counts (summed client-side).
  parts_count?: number | null;
  children: Subcategory[];
}

export interface PartsPage {
  items: import('./part').PublicPart[];
  total: number;
  page: number;
  pages: number;
  per_page: number;
}

export type PopularPartsPage = PartsPage;

/**
 * Faceted-search counts for the category page's filter UI.
 *
 * Each list is computed with every filter applied EXCEPT its own, so choosing
 * one manufacturer never collapses the manufacturer list to that single
 * option. `total_unfiltered` is the category's true size — the number the
 * "All" chip and the sub-sheet show — while `parts.total` is what the current
 * filters leave.
 */
export interface CategoryFacets {
  total_unfiltered: number;
  manufacturers: { name: string; count: number }[];
  /** SLUG-keyed. Empty on a leaf category (nothing to group by). */
  subs: { slug: string; name: string; count: number }[];
}

/**
 * Everything on a category page that does NOT change when you sort, filter,
 * search or paginate: identity, the family tree, and the sponsor boards.
 *
 * Split out from the rows so the two can be cached independently — the chrome
 * by slug (paints the header/chips on the first frame of a revisit), the rows
 * by slug + query signature (so a `?p=5&sort=qty100` URL can never paint
 * page-1 rows).
 */
export interface CategoryChrome extends Category {
  // Parent carries its own children (siblings of `this`) so subcategory pages
  // can render the SubcategoryChips strip without an extra fetch — see
  // 2026-05-16 fix for intra-category navigation on leaf pages.
  parent: {
    id: string;
    name: string;
    slug: string;
    icon: string;
    children: Subcategory[];
  } | null;
  // On a child category this is the single GOLD sponsor (newest visible); the
  // Gold-tier SponsorBlock consumes it. Null when unsold.
  sponsor: import('./sponsor').Sponsor | null;
  // The child category's SILVER sponsors (many) — feeds the SilverPartners
  // directory in the tier row. Empty on parent pages and when unsold.
  silver: import('./sponsor').PartnerSupplier[];
}

export interface CategoryDetail extends CategoryChrome {
  // ONE scope-aware block: a leaf's own parts, a parent's rollup over
  // self + children (each item still carries sub_slug + category names).
  // `total` is the FILTERED total — the header count, and the number the page
  // used to lie about while it truncated at 500 rows client-side.
  parts: PartsPage;
  // Legacy rollup block, unchanged and no longer read by this page: the page
  // asks for popular_per_page=1 and takes its parent ordering from `parts`
  // (sort=popular) instead. Kept because test_category_hierarchy pins it.
  popular_parts: PopularPartsPage;
  // Optional so a frontend deployed ahead of the API degrades (no counts, no
  // filter options) instead of white-screening. `?:` misses null, hence the
  // explicit `| null` — read it with `!= null`.
  facets?: CategoryFacets | null;
}

export interface CategoryPartners {
  // The resolved TOP-LEVEL category (a child slug resolves to its parent) and
  // its single PLATINUM Category Sponsor — fed to the always-present banner.
  // `platinum` is null when the slot is unsold (→ Open-Placement board).
  slug: string;
  name: string;
  platinum: import('./sponsor').PlatinumSponsor | null;
}
